### PR TITLE
[Feature] Support Distributed LogProb for GRPO Training

### PR DESCRIPTION
**Pull request by duanjunwen**
## 📌 Checklist before creating the PR

- [ ] I have created an issue for this PR for traceability
- [ ] The title follows the standard format: `[doc/gemini/tensor/...]: A concise description`
- [ ] I have added relevant tags if possible for us to better distinguish different PRs
- [ ] I have installed pre-commit: `pip install pre-commit && pre-commit install`

## 🚨 Issue number

> Link this PR to your issue with words like fixed to automatically close the linked issue upon merge
>
> e.g. `fixed #1234`, `closed #1234`, `resolved #1234`

## 📝 What does this PR do?

> Summarize your work here.
> if you have any plots/diagrams/screenshots/tables, please attach them here.

## 💥 Checklist before requesting a review

- [ ] I have linked my PR to an issue ([instruction](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue))
- [ ] My issue clearly describes the problem/feature/proposal, with diagrams/charts/table/code if possible
- [ ] I have performed a self-review of my code
- [ ] I have added thorough tests.
- [ ] I have added docstrings for all the functions/methods I implemented

## ⭐️ Do you enjoy contributing to Colossal-AI?

- [ ] 🌝 Yes, I do.
- [ ] 🌚 No, I don't.

Tell us more if you don't enjoy contributing to Colossal-AI.

### CODE REVIEW

File: colossalai/shardformer/layer/loss.py [hunk 1]
```diff
@@ -137,6 +144,87 @@ def backward(ctx, grad_output):
         return grad_logits, None, None, None, None, None, None
 
 
+class DistLogProb(Function):
+    r"""
+    Overwrite the forward and backward function to calculate the cross entropy loss before gather
```
Comment (TongLi3701) [inline colossalai/shardformer/layer/loss.py:149]: calculate log prob.

File: colossalai/shardformer/layer/loss.py [hunk 2]
```diff
@@ -137,6 +144,87 @@ def backward(ctx, grad_output):
...
+        ######
+        log_probs = log_probs.gather(dim=-1, index=masked_target.unsqueeze(-1))
+        print(f"log_probs in ops {log_probs.shape}")
+        # set masked val to zero, then all reduce
+        log_probs[mask.unsqueeze(-1)] = 0
+        # allreduce log_probs with ops SUM
+        dist.all_reduce(log_probs, op=dist.ReduceOp.SUM, group=process_group)
+        ctx.save_for_backward(exp_logits, mask, masked_target_1d, sum_exp_logits)
+        ctx.dtype = dtype
+        return log_probs
```
Comment (TongLi3701) [inline colossalai/shardformer/layer/loss.py:214]: Please remove unnecessary comments. For example, `allreduce log_probs with ops SUM` or  `cal exp_logits`, it is not necessary to explain this simple sentence. You should explain you algorithm (how to do forward here).

File: colossalai/shardformer/layer/loss.py [hunk 3]
```diff
@@ -137,6 +144,87 @@ def backward(ctx, grad_output):
...
+        # all_reduce get global sum_exp_logits
+        dist.all_reduce(sum_exp_logits, op=dist.ReduceOp.SUM, group=process_group)
+        # cal log_softmax
+        log_probs = vocab_logits - torch.log(sum_exp_logits.unsqueeze(dim=-1))
+
+        ######
+        # 2.gather via labels
+        ######
+        log_probs = log_probs.gather(dim=-1, index=masked_target.unsqueeze(-1))
+        print(f"log_probs in ops {log_probs.shape}")
```
Comment (TongLi3701) [inline colossalai/shardformer/layer/loss.py:207]: remove print.

File: colossalai/shardformer/layer/loss.py [hunk 4]
```diff
@@ -243,3 +341,51 @@ def dist_cross_entropy(
...
+    """
+    Helper to compute cross entropy loss for most shardformer models supporting PP, TP and SP.
+    """
+    # Split labels if not gather output
+    sp_group = shard_config.sequence_parallel_process_group
+    dist.get_rank(sp_group)
+    sp_size = shard_config.sequence_parallel_size
+    sp_mode = shard_config.sequence_parallelism_mode
+    parallel_output = shard_config.parallel_output
+    is_tp = shard_config.enable_tensor_parallelism
```
Comment (TongLi3701) [inline colossalai/shardformer/layer/loss.py:363]: we do not need sp related here.

File: colossalai/shardformer/layer/loss.py [hunk 5]
```diff
@@ -243,3 +341,51 @@ def dist_cross_entropy(
...
+
+    # Shift labels to predict the next token, and remove the tail logit predicting <EOS>
+    sp_size > 1 and (not is_share_sp_tp(sp_mode))
```
Comment (TongLi3701) [inline colossalai/shardformer/layer/loss.py:366]: not used.

**Review by TongLi3701 — CHANGES_REQUESTED** [review]
Thanks Junwen, I left some comments.

**Comment by TongLi3701** [comment]
Please also compare the peak memory when using these two different methods.

File: applications/ColossalChat/coati/distributed/consumer.py [hunk 6]
```diff
@@ -66,13 +66,15 @@ def setup(self) -> None:
         launch(self.rank, self.world_size, self.master_addr, self.master_port, local_rank=0)
 
         plugin_config = dict(
-            tp_size=1,
+            tp_size=2,
```
[the remark this answers is not in the document]
Comment (duanjunwen) [reply applications/ColossalChat/coati/distributed/consumer.py]: > Keep tp=1 as default.

Fixed.

File: applications/ColossalChat/coati/distributed/grpo_consumer.py [hunk 7]
```diff
@@ -116,18 +115,23 @@ def step(self, step_idx: int, **kwargs) -> Optional[float]:
 
         ctx = nullcontext() if need_update else self.booster.no_sync(self.policy_model, self.optimizer)
         with ctx:
-            policy_model_logits = self.policy_model(
+            policy_model_log_probs = self.policy_model(
                 input_ids=data["input_ids"],
                 attention_mask=data["attention_mask"],
+                return_dist_log_prob=True,
```
[the remark this answers is not in the document]
Comment (duanjunwen) [reply applications/ColossalChat/coati/distributed/grpo_consumer.py]: > why do you need a parameter for dist logprob? This should be automatically detected.

Removed

File: applications/ColossalChat/coati/distributed/grpo_consumer.py [hunk 8]
```diff
@@ -116,18 +115,23 @@ def step(self, step_idx: int, **kwargs) -> Optional[float]:
...
             )["logits"]
-            action_log_probs = calc_action_log_probs(policy_model_logits, data["input_ids"], num_action)
-
+            policy_model_log_probs = policy_model_log_probs.reshape(8, -1)
```
[the remark this answers is not in the document]
Comment (duanjunwen) [reply applications/ColossalChat/coati/distributed/grpo_consumer.py]: > should not hard code.

remove reshape here.

File: applications/ColossalChat/coati/distributed/grpo_consumer.py [hunk 9]
```diff
@@ -116,18 +115,23 @@ def step(self, step_idx: int, **kwargs) -> Optional[float]:
...
+                reference_model_log_probs = self.reference_model(
                     input_ids=data["input_ids"],
                     attention_mask=data["attention_mask"],
+                    return_dist_log_prob=True,
                 )["logits"]
-            reference_action_log_probs = calc_action_log_probs(reference_model_logits, data["input_ids"], num_action)
+                reference_model_log_probs = reference_model_log_probs.reshape(8, -1)
+                reference_action_log_probs = reference_model_log_probs[:, -num_action:]
+            # reference_action_log_probs = calc_action_log_probs(reference_model_logits, data["input_ids"], num_action)
+            # reference_action_log_probs = calc_action_log_probs(reference_model_logits, data["input_ids"], num_action, self.booster.shard_config, reference_model_logits.shape[-1])
```
[the remark this answers is not in the document]
Comment (duanjunwen) [reply applications/ColossalChat/coati/distributed/grpo_consumer.py]: > Same here.

No more hard code

File: applications/ColossalChat/coati/distributed/utils.py [hunk 10]
```diff
@@ -66,18 +68,52 @@ def log_probs_from_logits(logits: torch.Tensor, labels: torch.Tensor) -> torch.T
...
+
+    # log_probs = dist_log_prob(sequences[:, 1:], logits[:, :-1, :], shard_config, vocab_size, logits.dtype)
+    # # labels: torch.Tensor,  # [B, S] or [B, S, Vocab_size]
+    # logits: torch.Tensor,  # [B, S, Vocab_size]
+    # shard_config: ShardConfig,
+    # vocab_size: int,
+    # dtype: torch.dtype,
+    # seq_dim: int = 1,
+    # log_probs = log_probs_from_logits(logits[:, :-1, :], sequences[:, 1:])
     return log_probs[:, -num_actions:]
```
[the remark this answers is not in the document]
Comment (duanjunwen) [reply applications/ColossalChat/coati/distributed/utils.py]: > This implementation will cause the non-dist version not working.

It will check tp and gather output in dist loss func. But fine, I add a tp check in calc_action_log_probs

File: tests/test_shardformer/test_layer/test_dist_log_prob.py [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+
+    pred.retain_grad()
+    loss.mean().backward()
+
+    dist_pred = pred.clone().chunk(world_size, -1)[rank].detach()
+    dist_pred.requires_grad = True
+    dist_loss = dist_log_prob_1d(dist_pred, labels)
+
+    dist_pred.retain_grad()
+    dist_loss.squeeze(-1).mean().backward()
```
[the remark this answers is not in the document]
Comment (duanjunwen) [reply tests/test_shardformer/test_layer/test_dist_log_prob.py]: > it is not loss here, we can say logprob here

Fix name in https://github.com/hpcaitech/ColossalAI/pull/6247/commits/188d69dccf185548af7e1dad53c727ffab9198e6

File: colossalai/shardformer/policies/qwen2.py [hunk 12]
```diff
@@ -444,9 +450,18 @@ def module_policy(self):
                     sub_module_replacement=[
                         SubModuleReplacementDescription(
                             suffix="lm_head",
-                            target_module=LinearWithGradAccum,
+                            target_module=PaddingLMHead,
```
[the remark this answers is not in the document]
Comment (duanjunwen) [reply colossalai/shardformer/policies/qwen2.py]: > why do we need to replace with `PaddingLMHead` here.

want to test memory change, forget to restore it

**Comment by duanjunwen** [comment]
> Please also compare the peak memory when using these two different methods.

We compared the case where the parallel output is True or False under strategy tp2zero1.
When parallel output  is False, the peak mem is 93000+ MB. When parallel output is True, the peak mem is around 82000 MB.

File: colossalai/shardformer/layer/loss.py [hunk 13]
```diff
@@ -243,3 +341,51 @@ def dist_cross_entropy(
...
+        log_prob = dist_log_prob_1d(
+            logits,
+            labels,
+            process_group=shard_config.tensor_parallel_process_group,
+            vocab_size=vocab_size,
+            dtype=dtype,
+        )
+    else:
+        # NOTE if use TP and not parallel_output, the output is gathered in VocabParallelLMHead1D
+        logits = logits.view(-1, logits.size(-1))
```
[the remark this answers is not in the document]
Comment (duanjunwen) [reply colossalai/shardformer/layer/loss.py]: > Why do we need to view here?

Removed

File: colossalai/shardformer/layer/loss.py [hunk 14]
```diff
@@ -243,3 +341,51 @@ def dist_cross_entropy(
...
+        log_prob = loss_fct(logits)
```
[the remark this answers is not in the document]
Comment (duanjunwen) [reply colossalai/shardformer/layer/loss.py:387]: > non-dist version is not correct, you should also gather from labels.
> 
> Please check the original implementation as well.

Remove non-dist version in dist_log_prob. Now, we use non-dist log_prob from utils directly.

File: colossalai/shardformer/layer/loss.py [hunk 15]
```diff
@@ -243,3 +341,51 @@ def dist_cross_entropy(
...
+
+    # TODO:support sp
+    labels = labels[..., 1:]
+    logits = logits[..., :-1, :]
+    labels = labels.contiguous()
+    logits = logits.contiguous()
+    assert labels.shape == logits.shape[:-1], f"label shape {labels.shape} does not match logit shape {logits.shape}"
+
+    # Flatten the tokens
+    loss_fct = LogSoftmax()
```
[the remark this answers is not in the document]
Comment (duanjunwen) [reply colossalai/shardformer/layer/loss.py]: > remove `loss_fct` and use `LogSoftmax` directly.

Removed. Now, we use non-dist log_prob from utils directly.

**Review by TongLi3701 — COMMENTED** [review]
Thanks Junwen, I left some comments. Please address the comments and click on resolved then request review again.

Thanks.

**Comment by duanjunwen** [comment]
Resolve Conflict.

File: applications/ColossalChat/coati/distributed/consumer.py [hunk 16]
```diff
@@ -74,7 +74,7 @@ def setup(self) -> None:
         if self.plugin_config.get("pp_size", 1) > 1 and "num_microbatches" not in self.plugin_config:
             plugin_config["microbatch_size"] = self.microbatch_size
         if self.plugin_config.get("tp_size", 1) > 1:
-            plugin_config["parallel_output"] = False
+            plugin_config["parallel_output"] = True
```
Comment (TongLi3701) [inline applications/ColossalChat/coati/distributed/consumer.py:77]: This check can be removed as we implemented dist version.

File: colossalai/shardformer/layer/loss.py (hunk 1 above)
Comment (TongLi3701) [inline colossalai/shardformer/layer/loss.py:149]: calculate log prob

Comment (duanjunwen) [reply colossalai/shardformer/layer/loss.py:149]: > calculate log prob.

Fixed

File: colossalai/shardformer/layer/loss.py (hunk 3 above)
Comment (duanjunwen) [reply colossalai/shardformer/layer/loss.py:207]: > remove print.

Removed

File: colossalai/shardformer/layer/loss.py (hunk 5 above)
Comment (duanjunwen) [reply colossalai/shardformer/layer/loss.py:366]: > not used.

Remove the statement about sp

File: applications/ColossalChat/coati/distributed/utils.py [hunk 17]
```diff
@@ -66,18 +68,33 @@ def log_probs_from_logits(logits: torch.Tensor, labels: torch.Tensor) -> torch.T
...
         torch.Tensor: Action log probs.
     """
-    log_probs = log_probs_from_logits(logits[:, :-1, :], sequences[:, 1:])
+    # labels: torch.Tensor,  # [B, S] or [B, S, Vocab_size]
+    # logits: torch.Tensor,  # [B, S, Vocab_size]
+    if shard_config.tensor_parallel_size > 1 and shard_config.parallel_output:
+        log_probs = dist_log_prob(sequences, logits, shard_config, vocab_size, logits.dtype)
+        log_probs = log_probs.squeeze(-1)
+    else:
+        log_probs = log_probs_from_logits(logits[:, :-1, :], sequences[:, 1:])
```
[the remark this answers is not in the document]
Comment (duanjunwen) [reply applications/ColossalChat/coati/distributed/utils.py]: > `dist_log_prob` and `log_probs_from_logits ` are same level of function but different version, which means that one is for non-tp and one is for tp.
> 
> Thus, the input and output should keep as similar as possible.
> 
> For example, both functions have `log_probs = log_probs.squeeze(-1)`, but we have one inside the function but one outside the function, which is not a good practice.
Now we align tp and non-tp, both squeeze in dist_log_prob

File: colossalai/shardformer/layer/loss.py (hunk 2 above)
Comment (duanjunwen) [reply colossalai/shardformer/layer/loss.py:214]: > In other words, you need to explain firstly, we find the global maximum value of logits, then xxx.

Add a step-by-step comments

File: applications/ColossalChat/coati/distributed/consumer.py (hunk 16 above)
Comment (duanjunwen) [reply applications/ColossalChat/coati/distributed/consumer.py]: > This check can be removed as we implemented dist version.

Removed

File: colossalai/shardformer/layer/loss.py [hunk 18]
```diff
@@ -137,6 +144,98 @@ def backward(ctx, grad_output):
...
+        ##################
+        # Step3:Calculate global summation exp logits
+        ##################
+        vocab_logits = vocab_logits - logits_max.unsqueeze(dim=-1)
+        exp_logits = torch.exp(vocab_logits)
+        sum_exp_logits = torch.sum(exp_logits, dim=-1, dtype=torch.float32)  # local summation exp logits
+        dist.all_reduce(sum_exp_logits, op=dist.ReduceOp.SUM, group=process_group)
+
+        ##################
+        # Step4:Calculate local prob. We first cal log_softmax, them
```
Comment (TongLi3701) [inline colossalai/shardformer/layer/loss.py:205]: what is the aim of `them`.

File: colossalai/shardformer/layer/loss.py [hunk 19]
```diff
@@ -137,6 +144,98 @@ def backward(ctx, grad_output):
...
+        ##################
+        log_probs = vocab_logits - torch.log(sum_exp_logits.unsqueeze(dim=-1))  # cal log_softmax
+        log_probs = log_probs.gather(dim=-1, index=masked_target.unsqueeze(-1))
+        log_probs[mask.unsqueeze(-1)] = 0  # # set masked val to zero
```
Comment (TongLi3701) [inline colossalai/shardformer/layer/loss.py:209]: remove additional `#`

**Review by TongLi3701 — APPROVED** [review]

**Review by TongLi3701 — APPROVED** [review]
